lib/AST/DeclContext.cpp

```diff
@@ -969,3 +969,60 @@ Accessibility AccessScope::accessibilityForDiagnostics() const {
 
   return Accessibility::Private;
 }
+
+/// Return the DeclContext to compare when checking shared private access in
+/// Swift 4 mode. The shared scope is the type declaration if the context
+/// and the type declaration are in the same file, otherwise it is the types
+/// last extension in the source file. If the context does not refer to a
+/// declaration or extension, the supplied context is returned.
+static const DeclContext *
+getSharedPrivateDeclContext(const DeclContext *DC, const SourceFile *useSF) {
+  auto NTD = DC->getAsNominalTypeOrNominalTypeExtensionContext();
+  if (!NTD)
+    return DC;
+
+  // use the type declaration as the private scope if it is in the same
+  // file as useSF. This occurs for both extensions and declarations.
+  if (NTD->getParentSourceFile() == useSF)
+    return NTD;
+
+  // Otherwise use the last extension declaration in the same file.
+  const DeclContext *lastExtension = nullptr;
+  for (ExtensionDecl *ED : NTD->getExtensions())
+    if (ED->getParentSourceFile() == useSF)
+      lastExtension = ED;
+
+  // If there's no last extension, return the supplied context.
+  return lastExtension ?: DC;
```

Comment:
We try not to use GNU extensions like ?:, please just write lastExtension ? DC : nullptr